invenio_records_resources/services/records/queryparser/transformer.py

```diff
@@ -33,15 +33,57 @@ def term_name(self):
         """Get the term name."""
         return self._term_name
 
-    def map_word(self, node):
+    def map_word(self, node, **kwargs):
         """Modify a word node."""
         return self._word_fun(node) if self._word_fun else node
 
-    def map_phrase(self, node):
+    def map_phrase(self, node, **kwargs):
         """Modify a phrase node."""
         return self._phrase_fun(node) if self._phrase_fun else node
 
 
+class RestrictedTerm:
+    """Class used to apply specific permissions to search."""
+
+    def __init__(self, permission, term_name=None, word=None, phrase=None):
+        """Constructor."""
+        self.permission = permission
+        self._word_fun = word
+        self._phrase_fun = phrase
+        self._term_name = term_name
+
+    def allows(self, identity):
+        """Check the permission for the whole term."""
+        if not (self._word_fun or self._phrase_fun):
+            return self.permission.allows(identity)
+        else:
+            # if we specified the rewrite phrase of word function,
+            # let them take over the permission check
+            return True
+
+    @property
+    def term_name(self):
+        """Get the term name."""
+        return self._term_name
+
+    def map_word(self, node, context, **kwargs):
+        """Modify a word node."""
+        is_restricted = not self.permission.allows(context["identity"])
```

Comment:
we cannot move it to the constructor, since context is not available at the init time

Reply:
OK, but then, depending on the permissions evaluations, it could by cached by Identity to avoid re-evaluation.
If it is simply a `set()` `intersection`, then no need. However, is it will go through all the generators to have the final `set`, I would really evaluate caching it. To be checked?